File: Java/src/main/java/org/onlineoptimisation/problem/benchmark/BentCigarFunction.java
package org.onlineoptimisation.problem.benchmark;

import org.armadillojava.Arma;
import org.armadillojava.Col;
import org.armadillojava.Mat;

public class BentCigarFunction extends BenchmarkProblem {

  protected Mat _R;
  
  public BentCigarFunction(int numberOfDimensions) {
    super(numberOfDimensions);
    
    _R = randomRotation();
  }

  @Override
  public double getObjectiveValueImplementation(Col parameter) {
    Col z = Arma.square(_R.times(asymmetric(0.5, _R.times(parameter))));
    return  1000000 * Arma.accu(z) - 999999 * z.at(1);
  }

}

File: src/optimisationProblem/benchmark/bentCigarFunction.cpp
#include <optimisationProblem/benchmark/bentCigarFunction.hpp>

namespace hop {
  BentCigarFunction::BentCigarFunction(const unsigned int &numberOfDimensions) : BenchmarkProblem(numberOfDimensions) {

  }

  double BentCigarFunction::getObjectiveValueImplementation(const arma::Col<double> &parameter) const {
    arma::Col<double> z = arma::square(_rotationR * getAsymmetricTransformation(0.5, _rotationR * (parameter - _translation)));
    return z.at(0) + 1000000 * arma::accu(z.subvec(1, z.n_elem - 1));
  }
}
